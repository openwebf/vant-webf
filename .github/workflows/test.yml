name: CI

on:
  push:
    branches:
      - '**'

  pull_request:
    branches:
      - dev

  workflow_dispatch:

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Run linter
      run: npm run lint

  test:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Run test cases
      run: npm test

    - name: Upload coverage to Codecov  
      uses: codecov/codecov-action@v1
      with:
        token: ${{ secrets.CODECOV_TOKEN }}

  build:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Build
      run: npm run build
