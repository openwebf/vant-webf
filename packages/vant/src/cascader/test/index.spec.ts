import { Cascader } from '..';
import { mount, later } from '../../../test';
import options from '../demo/area-en-US';

test('should emit change event when active option changed', async () => {
  const wrapper = mount(Cascader, {
    props: {
      options,
    },
  });

  await later();
  wrapper.find('.van-cascader__option').trigger('click');

  const firstOption = options[0];
  expect(wrapper.emitted('change')![0]).toEqual([
    {
      value: firstOption.value,
      tabIndex: 0,
      selectedOptions: [firstOption],
    },
  ]);

  await later();
  wrapper
    .findAll('.van-cascader__options')[1]
    .find('.van-cascader__option')
    .trigger('click');
  const secondOption = options[0].children[0];
  expect(wrapper.emitted('change')![1]).toEqual([
    {
      value: secondOption.value,
      tabIndex: 1,
      selectedOptions: [firstOption, secondOption],
    },
  ]);
});

test('should emit finish event when all options is selected', async () => {
  const option = { value: '1', text: 'foo' };
  const wrapper = mount(Cascader, {
    props: {
      options: [option],
    },
  });

  await later();
  wrapper.find('.van-cascader__option').trigger('click');
  expect(wrapper.emitted('finish')![0]).toEqual([
    {
      value: option.value,
      tabIndex: 0,
      selectedOptions: [option],
    },
  ]);
});

test('should emit close event when close icon is clicked', () => {
  const wrapper = mount(Cascader);
  wrapper.find('.van-cascader__close-icon').trigger('click');
  expect(wrapper.emitted('close')![0]).toBeTruthy();
});

test('should not render close icon when closeable is false', () => {
  const wrapper = mount(Cascader, {
    props: {
      closeable: false,
    },
  });
  expect(wrapper.find('.van-cascader__close-icon').exists()).toBeFalsy();
});

test('should change close icon when using close-icon prop', () => {
  const wrapper = mount(Cascader, {
    props: {
      closeIcon: 'success',
    },
  });
  expect(wrapper.find('.van-cascader__close-icon').html()).toMatchSnapshot();
});

test('should render title slot correctly', () => {
  const wrapper = mount(Cascader, {
    slots: {
      title: () => 'Custom Title',
    },
  });
  expect(wrapper.find('.van-cascader__title').html()).toMatchSnapshot();
});

test('should render option slot correctly', async () => {
  const option = { value: '1', text: 'foo' };
  const wrapper = mount(Cascader, {
    props: {
      options: [option],
    },
    slots: {
      option: ({ option }) => `Custom Option ${option.text}`,
    },
  });
  await later();
  expect(wrapper.find('.van-cascader__option').html()).toMatchSnapshot();
});

test('should select correct option when value changed', async () => {
  const wrapper = mount(Cascader, {
    props: {
      options,
    },
  });

  await later();
  await wrapper.setProps({ modelValue: '330304' });
  await later();
  const selectedOptions = wrapper.findAll('.van-cascader__option--selected');
  const lastSelectedOption = selectedOptions[selectedOptions.length - 1];
  expect(lastSelectedOption.html()).toMatchSnapshot();
});

test('should reset selected options when value is set to empty', async () => {
  const wrapper = mount(Cascader, {
    props: {
      options,
      modelValue: '330304',
    },
  });

  await wrapper.setProps({ modelValue: '' });
  expect(wrapper.find('.van-cascader__option--selected').exists()).toBeFalsy();
});

test('should update tabs when previous tab is clicked', async () => {
  const wrapper = mount(Cascader, {
    props: {
      options,
      modelValue: '330304',
    },
  });

  await later();
  wrapper.findAll('.van-cascader__option')[1].trigger('click');
  await later();
  expect(wrapper.html()).toMatchSnapshot();
});

test('should allow to custom field names', async () => {
  const fieldNames = {
    text: 'name',
    value: 'code',
    children: 'items',
  };
  const options = [
    {
      name: 'Zhejiang',
      code: '330000',
      items: [{ name: 'Hangzhou', code: '330100' }],
    },
  ];
  const wrapper = mount(Cascader, {
    props: {
      options,
      fieldNames,
    },
  });

  await later();
  wrapper.find('.van-cascader__option').trigger('click');

  const firstOption = options[0];
  expect(wrapper.emitted('change')![0]).toEqual([
    {
      value: firstOption.code,
      tabIndex: 0,
      selectedOptions: [firstOption],
    },
  ]);

  await later();
  wrapper
    .findAll('.van-cascader__options')[1]
    .find('.van-cascader__option')
    .trigger('click');
  const secondOption = options[0].items[0];
  expect(wrapper.emitted('change')![1]).toEqual([
    {
      value: secondOption.code,
      tabIndex: 1,
      selectedOptions: [firstOption, secondOption],
    },
  ]);
});

test('should emit click-tab event when a tab is clicked', async () => {
  const wrapper = mount(Cascader, {
    props: {
      options,
    },
  });

  await later();
  wrapper.find('.van-cascader__option').trigger('click');
  await later();
  wrapper
    .findAll('.van-cascader__options')[1]
    .find('.van-cascader__option')
    .trigger('click');
  await later();

  const tabs = wrapper.findAll('.van-tab');

  tabs[0].trigger('click');
  expect(wrapper.emitted('click-tab')![0]).toEqual([0, options[0].text]);

  tabs[1].trigger('click');
  expect(wrapper.emitted('click-tab')![1]).toEqual([
    1,
    options[0].children[0].text,
  ]);
});

test('should allow to custom the className of option', async () => {
  const wrapper = mount(Cascader, {
    props: {
      options: [{ value: '1', text: 'foo', className: 'foo' }],
    },
  });

  await later();
  const option = wrapper.find('.van-cascader__option');
  expect(option.classes()).toContain('foo');
});

test('should allow to custom the color of option', async () => {
  const wrapper = mount(Cascader, {
    props: {
      options: [{ value: '1', text: 'foo', color: 'red' }],
    },
  });

  await later();
  const option = wrapper.find('.van-cascader__option');
  expect(option.style.color).toEqual('red');
});

test('should render options-top、options-bottom slots correctly', async () => {
  const wrapper = mount(Cascader, {
    slots: {
      'options-top': ({ tabIndex }) => `Top, tab index: ${tabIndex}`,
      'options-bottom': ({ tabIndex }) => `Bottom, tab index: ${tabIndex}`,
    },
    props: {
      options,
    },
  });

  await later();
  await wrapper
    .findAll('.van-cascader__options')[0]
    .find('.van-cascader__option')
    .trigger('click');

  expect(wrapper.find('.van-tab__panel').html()).toMatchSnapshot();
});

test('should not render header when show-header prop is false', async () => {
  const wrapper = mount(Cascader, {
    props: {
      options,
      showHeader: false,
    },
  });

  const header = wrapper.find('.van-cascader__header');
  expect(header.exists()).toBeFalsy();
});
