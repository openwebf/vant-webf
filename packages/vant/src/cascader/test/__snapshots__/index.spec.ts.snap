// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should change close icon when using close-icon prop 1`] = `
<i class="van-badge__wrapper van-icon van-icon-success van-cascader__close-icon van-haptics-feedback">
</i>
`;

exports[`should render option slot correctly 1`] = `
<li role="menuitemradio"
    class="van-cascader__option"
    tabindex="-1"
    aria-checked="false"
>
  Custom Option foo
</li>
`;

exports[`should render options-top„ÄÅoptions-bottom slots correctly 1`] = `
<div class="van-tab__panel">
  Top, tab index: 0
  <ul role="menu"
      class="van-cascader__options"
  >
    <li role="menuitemradio"
        class="van-cascader__option van-cascader__option--selected"
        tabindex="0"
        aria-checked="true"
    >
      <span>
        Zhejiang
      </span>
      <i class="van-badge__wrapper van-icon van-icon-success van-cascader__selected-icon">
      </i>
    </li>
    <li role="menuitemradio"
        class="van-cascader__option"
        tabindex="-1"
        aria-checked="false"
    >
      <span>
        Jiangsu
      </span>
    </li>
  </ul>
  Bottom, tab index: 0
</div>
`;

exports[`should render title slot correctly 1`] = `
<h2 class="van-cascader__title">
  Custom Title
</h2>
`;

exports[`should select correct option when value changed 1`] = `
<li role="menuitemradio"
    class="van-cascader__option van-cascader__option--selected"
    tabindex="0"
    aria-checked="true"
>
  <span>
    Wenzhou
  </span>
  <i class="van-badge__wrapper van-icon van-icon-success van-cascader__selected-icon">
  </i>
</li>
`;

exports[`should update tabs when previous tab is clicked 1`] = `
<div class="van-cascader">
  <div class="van-cascader__header">
    <h2 class="van-cascader__title">
    </h2>
    <i class="van-badge__wrapper van-icon van-icon-cross van-cascader__close-icon van-haptics-feedback">
    </i>
  </div>
  <div class="van-tabs van-tabs--line van-cascader__tabs">
    <div class="van-tabs__wrap">
      <div role="tablist"
           class="van-tabs__nav van-tabs__nav--line van-tabs__nav--shrink van-tabs__nav--complete"
           aria-orientation="horizontal"
      >
        <div id="van-tabs-0"
             role="tab"
             class="van-tab van-tab--line van-tab--shrink van-cascader__tab"
             tabindex="-1"
             aria-selected="false"
        >
          <span class="van-tab__text">
            Jiangsu
          </span>
        </div>
        <div id="van-tabs-1"
             role="tab"
             class="van-tab van-tab--line van-tab--shrink van-tab--active van-cascader__tab van-cascader__tab--unselected"
             tabindex="0"
             aria-selected="true"
        >
          <span class="van-tab__text">
            Select
          </span>
        </div>
        <div class="van-tabs__line"
             style="transform: translateX(50px) translateX(-50%); transition-duration: 0.3s;"
        >
        </div>
      </div>
    </div>
    <div class="van-tabs__content van-tabs__content--animated">
      <div class="van-swipe van-tabs__track">
        <div style="transition-duration: 0ms; transform: translateX(0px); width: 200px;"
             class="van-swipe__track"
        >
          <div class="van-swipe-item van-tab__panel-wrapper"
               id="van-tab"
               role="tabpanel"
               tabindex="-1"
               aria-hidden="true"
               aria-labelledby="van-tabs-0"
               style="width: 100px;"
          >
            <div class="van-tab__panel">
              <ul role="menu"
                  class="van-cascader__options"
              >
                <li role="menuitemradio"
                    class="van-cascader__option"
                    tabindex="-1"
                    aria-checked="false"
                >
                  <span>
                    Zhejiang
                  </span>
                </li>
                <li role="menuitemradio"
                    class="van-cascader__option van-cascader__option--selected"
                    tabindex="0"
                    aria-checked="true"
                >
                  <span>
                    Jiangsu
                  </span>
                  <i class="van-badge__wrapper van-icon van-icon-success van-cascader__selected-icon">
                  </i>
                </li>
              </ul>
            </div>
          </div>
          <div class="van-swipe-item van-tab__panel-wrapper"
               id="van-tab"
               role="tabpanel"
               tabindex="0"
               aria-hidden="false"
               aria-labelledby="van-tabs-1"
               style="width: 100px;"
          >
            <div class="van-tab__panel">
              <ul role="menu"
                  class="van-cascader__options"
              >
                <li role="menuitemradio"
                    class="van-cascader__option"
                    tabindex="-1"
                    aria-checked="false"
                >
                  <span>
                    Nanjing
                  </span>
                </li>
                <li role="menuitemradio"
                    class="van-cascader__option"
                    tabindex="-1"
                    aria-checked="false"
                >
                  <span>
                    Wuxi
                  </span>
                </li>
                <li role="menuitemradio"
                    class="van-cascader__option"
                    tabindex="-1"
                    aria-checked="false"
                >
                  <span>
                    Xuzhou
                  </span>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
